File: client/cs/src/Proton.Drive.Sdk.CExports/InteropProtonDriveClient.cs
using System.Text.Json;
using Google.Protobuf;
using Google.Protobuf.WellKnownTypes;
using Proton.Drive.Sdk.Nodes;
using Proton.Drive.Sdk.Nodes.Download;
using Proton.Drive.Sdk.Nodes.Upload;
using Proton.Sdk.Caching;
using Proton.Sdk.Configuration;
using Proton.Sdk.Telemetry;

namespace Proton.Drive.Sdk.CExports;

internal static class InteropProtonDriveClient
{
    public static IMessage HandleCreate(DriveClientCreateRequest request, nint bindingsHandle)
    {
        if (!request.BaseUrl.EndsWith('/'))
        {
            throw new UriFormatException("Base URL must end with a '/'");
        }

        var protonDriveClientOptions = new Sdk.ProtonDriveClientOptions(
            request.ClientOptions.HasUid ? request.ClientOptions.Uid : null,
            request.ClientOptions.HasBindingsLanguage ? request.ClientOptions.BindingsLanguage : null,
            request.ClientOptions.HasApiCallTimeout ? request.ClientOptions.ApiCallTimeout : null,
            request.ClientOptions.HasStorageCallTimeout ? request.ClientOptions.StorageCallTimeout : null,
            request.ClientOptions.HasBlockTransferParallelism ? request.ClientOptions.BlockTransferParallelism : null);

        var httpClientFactory = new InteropHttpClientFactory(
            bindingsHandle,
            request.BaseUrl,
            protonDriveClientOptions.BindingsLanguage,
            new InteropFunction<nint, InteropArray<byte>, nint, nint>(request.HttpClient.RequestFunction),
            new InteropFunction<nint, InteropArray<byte>, nint, nint>(request.HttpClient.ResponseContentReadAction),
            new InteropAction<nint>(request.HttpClient.ResponseContentDisposeAction),
            new InteropAction<nint>(request.HttpClient.CancellationAction));

        var accountClient = new InteropProtonAccountClient(bindingsHandle, new InteropAction<nint, InteropArray<byte>, nint>(request.AccountRequestAction));

        ICacheRepository? cacheRepository = request.HasCachePath
            ? SqliteCacheRepository.OpenFile(request.CachePath)
            : null;

        if (request.HasCacheEncryptionKey && cacheRepository is not null)
        {
            cacheRepository = new EncryptedCacheRepository(cacheRepository, request.CacheEncryptionKey.ToByteArray());
        }

        ITelemetry telemetry = request.Telemetry.ToTelemetry(bindingsHandle) is { } interopTelemetry
            ? new DriveInteropTelemetryDecorator(interopTelemetry)
            : NullTelemetry.Instance;

        var featureFlagProvider = request.HasFeatureEnabledFunction
            ? new InteropFeatureFlagProvider(bindingsHandle, new InteropFunction<nint, InteropArray<byte>, int>(request.FeatureEnabledFunction))
            : AlwaysDisabledFeatureFlagProvider.Instance;

        var client = new ProtonDriveClient(
            httpClientFactory,
            accountClient,
            cacheRepository,
            featureFlagProvider,
            telemetry,
            protonDriveClientOptions);

        return new Int64Value
        {
            Value = Interop.AllocHandle(client),
        };
    }

    public static async ValueTask<IMessage> HandleCreateFolderAsync(DriveClientCreateFolderRequest request)
    {
        var cancellationToken = Interop.GetCancellationToken(request.CancellationTokenSourceHandle);

        var client = Interop.GetFromHandle<ProtonDriveClient>(request.ClientHandle);

        var createdFolder = await client.CreateFolderAsync(
            NodeUid.Parse(request.ParentFolderUid),
            request.FolderName,
            request.LastModificationTime?.ToDateTimeFixed(),
            cancellationToken).ConfigureAwait(false);

        return createdFolder.ToInterop();
    }

    public static async ValueTask<IMessage> HandleGetFileUploaderAsync(DriveClientGetFileUploaderRequest request)
    {
        var cancellationToken = Interop.GetCancellationToken(request.CancellationTokenSourceHandle);
        var additionalMetadata = request.AdditionalMetadata is { Count: > 0 }
            ? request.AdditionalMetadata.Select(x =>
                new Nodes.AdditionalMetadataProperty(x.Name, JsonDocument.Parse(x.Utf8JsonValue.Memory).RootElement))
            : null;

        var metadata = new FileUploadMetadata
        {
            LastModificationTime = request.LastModificationTime?.ToDateTimeFixed(),
            AdditionalMetadata = additionalMetadata,
        };

        var client = Interop.GetFromHandle<ProtonDriveClient>(request.ClientHandle);

        FileUploader? fileUploader;
        if (request is { HasNoWaiting: true, NoWaiting: true })
        {
#pragma warning disable TryTransferQueuing
            fileUploader = client.TryGetFileUploader(
                NodeUid.Parse(request.ParentFolderUid),
                request.Name,
                request.MediaType,
                request.Size,
                metadata,
                request.OverrideExistingDraftByOtherClient);
#pragma warning restore TryTransferQueuing
        }
        else
        {
            fileUploader = await client.GetFileUploaderAsync(
                NodeUid.Parse(request.ParentFolderUid),
                request.Name,
                request.MediaType,
                request.Size,
                metadata,
                request.OverrideExistingDraftByOtherClient,
                cancellationToken).ConfigureAwait(false);
        }

        return new Int64Value { Value = fileUploader is null ? 0 : Interop.AllocHandle(fileUploader) };
    }

    public static async ValueTask<IMessage> HandleGetFileRevisionUploaderAsync(DriveClientGetFileRevisionUploaderRequest request)
    {
        var cancellationToken = Interop.GetCancellationToken(request.CancellationTokenSourceHandle);

        var client = Interop.GetFromHandle<ProtonDriveClient>(request.ClientHandle);

        var additionalMetadata = request.AdditionalMetadata.Count > 0
            ? request.AdditionalMetadata.Select(x =>
                new Nodes.AdditionalMetadataProperty(x.Name, JsonDocument.Parse(x.Utf8JsonValue.Memory).RootElement))
            : null;

        var metadata = new FileUploadMetadata
        {
            LastModificationTime = request.LastModificationTime?.ToDateTimeFixed(),
            AdditionalMetadata = additionalMetadata,
        };

        FileUploader? fileUploader;
        if (request is { HasNoWaiting: true, NoWaiting: true })
        {
#pragma warning disable TryTransferQueuing
            fileUploader = client.TryGetFileRevisionUploader(
                RevisionUid.Parse(request.CurrentActiveRevisionUid),
                request.Size,
                metadata);
#pragma warning restore TryTransferQueuing
        }
        else
        {
            fileUploader = await client.GetFileRevisionUploaderAsync(
            RevisionUid.Parse(request.CurrentActiveRevisionUid),
            request.Size,
            metadata,
            cancellationToken).ConfigureAwait(false);
        }

        return new Int64Value { Value = fileUploader is null ? 0 : Interop.AllocHandle(fileUploader) };
    }

    public static async ValueTask<IMessage> HandleGetAvailableNameAsync(DriveClientGetAvailableNameRequest request)
    {
        var cancellationToken = Interop.GetCancellationToken(request.CancellationTokenSourceHandle);

        var client = Interop.GetFromHandle<ProtonDriveClient>(request.ClientHandle);

        var availableName = await client.GetAvailableNameAsync(
            NodeUid.Parse(request.ParentFolderUid),
            request.Name,
            cancellationToken).ConfigureAwait(false);

        return new StringValue { Value = availableName };
    }

    public static async ValueTask<IMessage?> HandleEnumerateThumbnailsAsync(DriveClientEnumerateThumbnailsRequest request, nint bindingsHandle)
    {
        var yieldAction = new InteropAction<nint, InteropArray<byte>>(request.YieldAction);
        var cancellationToken = Interop.GetCancellationToken(request.CancellationTokenSourceHandle);

        var client = Interop.GetFromHandle<ProtonDriveClient>(request.ClientHandle);

        var thumbnailsEnumerable = client.EnumerateThumbnailsAsync(
            request.FileUids.Select(NodeUid.Parse),
            (Nodes.ThumbnailType)request.Type,
            cancellationToken);

        await foreach (var x in thumbnailsEnumerable.ConfigureAwait(false))
        {
            var thumbnail = new FileThumbnail { FileUid = x.FileUid.ToString() };

            if (x.Result.TryGetValueElseError(out var data, out var error))
            {
                thumbnail.Data = ByteString.CopyFrom(data.Span);
            }
            else
            {
                thumbnail.Error = error.ToInterop();
            }

            yieldAction.InvokeWithMessage(bindingsHandle, thumbnail);
        }

        return null;
    }

    public static async ValueTask<IMessage?> HandleEnumerateFolderChildrenAsync(DriveClientEnumerateFolderChildrenRequest request, nint bindingsHandle)
    {
        var yieldAction = new InteropAction<nint, InteropArray<byte>>(request.YieldAction);
        var cancellationToken = Interop.GetCancellationToken(request.CancellationTokenSourceHandle);

        var client = Interop.GetFromHandle<ProtonDriveClient>(request.ClientHandle);

        await foreach (var nodeUid in client.EnumerateFolderChildrenNodeUidsAsync(NodeUid.Parse(request.FolderUid), cancellationToken).ConfigureAwait(false))
        {
            yieldAction.InvokeWithMessage(bindingsHandle, new StringValue { Value = nodeUid.ToString() });
        }

        return null;
    }

    public static async ValueTask<IMessage> HandleGetMyFilesFolderAsync(DriveClientGetMyFilesFolderRequest request)
    {
        var cancellationToken = Interop.GetCancellationToken(request.CancellationTokenSourceHandle);
        var client = Interop.GetFromHandle<ProtonDriveClient>(request.ClientHandle);

        var folderNode = await client.GetMyFilesFolderAsync(cancellationToken).ConfigureAwait(false);

        return folderNode.ToInterop();
    }

    public static async ValueTask<IMessage> HandleGetFileDownloaderAsync(DriveClientGetFileDownloaderRequest request)
    {
        var cancellationToken = Interop.GetCancellationToken(request.CancellationTokenSourceHandle);
        var client = Interop.GetFromHandle<ProtonDriveClient>(request.ClientHandle);
        var revisionUid = RevisionUid.Parse(request.RevisionUid);

        FileDownloader? fileDownloader;
        if (request is { HasNoWaiting: true, NoWaiting: true })
        {
#pragma warning disable TryTransferQueuing
            fileDownloader = client.TryGetFileDownloader(revisionUid);
#pragma warning restore TryTransferQueuing
        }
        else
        {
            fileDownloader = await client.GetFileDownloaderAsync(revisionUid, cancellationToken).ConfigureAwait(false);
        }

        return new Int64Value { Value = fileDownloader is null ? 0 : Interop.AllocHandle(fileDownloader) };
    }

    public static async ValueTask<IMessage?> HandleRenameAsync(DriveClientRenameRequest request)
    {
        var cancellationToken = Interop.GetCancellationToken(request.CancellationTokenSourceHandle);

        var client = Interop.GetFromHandle<ProtonDriveClient>(request.ClientHandle);

        await client.RenameNodeAsync(
            NodeUid.Parse(request.NodeUid),
            request.NewName,
            request.HasNewMediaType ? request.NewMediaType : null,
            cancellationToken).ConfigureAwait(false);

        return null;
    }

    public static async ValueTask<IMessage?> HandleTrashNodesAsync(DriveClientTrashNodesRequest request, nint bindingsHandle)
    {
        var yieldAction = new InteropAction<nint, InteropArray<byte>>(request.YieldAction);
        var cancellationToken = Interop.GetCancellationToken(request.CancellationTokenSourceHandle);

        var client = Interop.GetFromHandle<ProtonDriveClient>(request.ClientHandle);

        var results = client.TrashNodesAsync(request.NodeUids.Select(NodeUid.Parse), cancellationToken);

        await foreach (var result in results.ConfigureAwait(false))
        {
            yieldAction.InvokeWithMessage(bindingsHandle, result.ToInterop());
        }

        return null;
    }

    public static async ValueTask<IMessage> HandleMoveNodesAsync(DriveClientMoveNodesRequest request)
    {
        var cancellationToken = Interop.GetCancellationToken(request.CancellationTokenSourceHandle);

        var client = Interop.GetFromHandle<ProtonDriveClient>(request.ClientHandle);

        var results = await client.MoveNodesAsync(
            request.NodeUids.Select(NodeUid.Parse),
            NodeUid.Parse(request.NewParentFolderUid),
            cancellationToken).ConfigureAwait(false);

        return results.ToInterop();
    }

    public static async ValueTask<IMessage?> HandleDeleteNodesAsync(DriveClientDeleteNodesRequest request, nint bindingsHandle)
    {
        var yieldAction = new InteropAction<nint, InteropArray<byte>>(request.YieldAction);
        var cancellationToken = Interop.GetCancellationToken(request.CancellationTokenSourceHandle);

        var client = Interop.GetFromHandle<ProtonDriveClient>(request.ClientHandle);

        var results = client.DeleteNodesAsync(request.NodeUids.Select(NodeUid.Parse), cancellationToken);

        await foreach (var result in results.ConfigureAwait(false))
        {
            yieldAction.InvokeWithMessage(bindingsHandle, result.ToInterop());
        }

        return null;
    }

    public static async ValueTask<IMessage?> HandleRestoreNodesAsync(DriveClientRestoreNodesRequest request, nint bindingsHandle)
    {
        var yieldAction = new InteropAction<nint, InteropArray<byte>>(request.YieldAction);
        var cancellationToken = Interop.GetCancellationToken(request.CancellationTokenSourceHandle);

        var client = Interop.GetFromHandle<ProtonDriveClient>(request.ClientHandle);

        var results = client.RestoreNodesAsync(request.NodeUids.Select(NodeUid.Parse), cancellationToken);

        await foreach (var result in results.ConfigureAwait(false))
        {
            yieldAction.InvokeWithMessage(bindingsHandle, result.ToInterop());
        }

        return null;
    }

    public static async ValueTask<IMessage?> HandleEnumerateTrashAsync(DriveClientEnumerateTrashRequest request, nint bindingsHandle)
    {
        var yieldAction = new InteropAction<nint, InteropArray<byte>>(request.YieldAction);
        var cancellationToken = Interop.GetCancellationToken(request.CancellationTokenSourceHandle);

        var client = Interop.GetFromHandle<ProtonDriveClient>(request.ClientHandle);

        await foreach (var nodeUid in client.EnumerateTrashNodeUidsAsync(cancellationToken).ConfigureAwait(false))
        {
            yieldAction.InvokeWithMessage(bindingsHandle, new StringValue { Value = nodeUid.ToString() });
        }

        return null;
    }

    public static async ValueTask<IMessage?> HandleGetNodeAsync(DriveClientGetNodeRequest request)
    {
        var cancellationToken = Interop.GetCancellationToken(request.CancellationTokenSourceHandle);
        var client = Interop.GetFromHandle<ProtonDriveClient>(request.ClientHandle);

        var node = await client.GetNodeAsync(
            NodeUid.Parse(request.NodeUid),
            cancellationToken).ConfigureAwait(false);

        return node?.ToInterop();
    }

    public static async ValueTask<IMessage?> HandleEnumerateSharedWithMeNodeUidsAsync(
        DriveClientEnumerateSharedWithMeNodeUidsRequest request,
        nint bindingsHandle)
    {
        var yieldAction = new InteropAction<nint, InteropArray<byte>>(request.YieldAction);
        var cancellationToken = Interop.GetCancellationToken(request.CancellationTokenSourceHandle);

        var client = Interop.GetFromHandle<ProtonDriveClient>(request.ClientHandle);

        await foreach (var nodeUid in client.EnumerateSharedWithMeNodeUidsAsync(cancellationToken).ConfigureAwait(false))
        {
            yieldAction.InvokeWithMessage(bindingsHandle, new StringValue { Value = nodeUid.ToString() });
        }

        return null;
    }

    public static async ValueTask<IMessage?> HandleEnumerateSharedNodeUidsAsync(
        DriveClientEnumerateSharedNodeUidsRequest request,
        nint bindingsHandle)
    {
        var yieldAction = new InteropAction<nint, InteropArray<byte>>(request.YieldAction);
        var cancellationToken = Interop.GetCancellationToken(request.CancellationTokenSourceHandle);

        var client = Interop.GetFromHandle<ProtonDriveClient>(request.ClientHandle);

        await foreach (var nodeUid in client.EnumerateSharedNodeUidsAsync(cancellationToken).ConfigureAwait(false))
        {
            yieldAction.InvokeWithMessage(bindingsHandle, new StringValue { Value = nodeUid.ToString() });
        }

        return null;
    }

    public static async ValueTask<IMessage?> HandleEnumerateEventsAsync(DriveClientEnumerateEventsRequest request, nint bindingsHandle)
    {
        var yieldAction = new InteropAction<nint, InteropArray<byte>>(request.YieldAction);
        var cancellationToken = Interop.GetCancellationToken(request.CancellationTokenSourceHandle);

        var client = Interop.GetFromHandle<ProtonDriveClient>(request.ClientHandle);

        var eventScopeId = new Events.DriveEventScopeId(new Volumes.VolumeId(request.TreeEventScopeId));
        Events.DriveEventId? cursorEventId = request.HasCursorEventId
            ? (Events.DriveEventId)request.CursorEventId
            : null;

        await foreach (var driveEvent in client.EnumerateEventsAsync(eventScopeId, cursorEventId, cancellationToken).ConfigureAwait(false))
        {
            yieldAction.InvokeWithMessage(bindingsHandle, driveEvent.ToInterop());
        }

        return null;
    }

    public static async ValueTask<IMessage?> HandleLeaveSharedNodeAsync(DriveClientLeaveSharedNodeRequest request)
    {
        var cancellationToken = Interop.GetCancellationToken(request.CancellationTokenSourceHandle);

        var client = Interop.GetFromHandle<ProtonDriveClient>(request.ClientHandle);

        await client.LeaveSharedNodeAsync(NodeUid.Parse(request.NodeUid), cancellationToken).ConfigureAwait(false);

        return null;
    }

    public static async ValueTask<IMessage?> HandleEmptyTrashAsync(DriveClientEmptyTrashRequest request)
    {
        var cancellationToken = Interop.GetCancellationToken(request.CancellationTokenSourceHandle);

        var client = Interop.GetFromHandle<ProtonDriveClient>(request.ClientHandle);

        await client.EmptyTrashAsync(cancellationToken).ConfigureAwait(false);

        return null;
    }

    public static async ValueTask<IMessage?> HandleEnumerateDevicesAsync(DriveClientEnumerateDevicesRequest request, nint bindingsHandle)
    {
        var yieldAction = new InteropAction<nint, InteropArray<byte>>(request.YieldAction);
        var cancellationToken = Interop.GetCancellationToken(request.CancellationTokenSourceHandle);

        var client = Interop.GetFromHandle<ProtonDriveClient>(request.ClientHandle);

        await foreach (var device in client.EnumerateDevicesAsync(cancellationToken).ConfigureAwait(false))
        {
            yieldAction.InvokeWithMessage(bindingsHandle, device.ToInterop());
        }

        return null;
    }

    public static async ValueTask<IMessage> HandleCreateDeviceAsync(DriveClientCreateDeviceRequest request)
    {
        var cancellationToken = Interop.GetCancellationToken(request.CancellationTokenSourceHandle);
        var client = Interop.GetFromHandle<ProtonDriveClient>(request.ClientHandle);

        var device = await client.CreateDeviceAsync(
            request.Name,
            (Devices.DeviceType)(int)request.DeviceType,
            cancellationToken).ConfigureAwait(false);

        return device.ToInterop();
    }

    public static async ValueTask<IMessage> HandleRenameDeviceAsync(DriveClientRenameDeviceRequest request)
    {
        var cancellationToken = Interop.GetCancellationToken(request.CancellationTokenSourceHandle);
        var client = Interop.GetFromHandle<ProtonDriveClient>(request.ClientHandle);

        var device = await client.RenameDeviceAsync(
            Devices.DeviceUid.Parse(request.DeviceUid),
            request.Name,
            cancellationToken).ConfigureAwait(false);

        return device.ToInterop();
    }

    public static async ValueTask<IMessage?> HandleDeleteDeviceAsync(DriveClientDeleteDeviceRequest request)
    {
        var cancellationToken = Interop.GetCancellationToken(request.CancellationTokenSourceHandle);
        var client = Interop.GetFromHandle<ProtonDriveClient>(request.ClientHandle);

        await client.DeleteDeviceAsync(Devices.DeviceUid.Parse(request.DeviceUid), cancellationToken).ConfigureAwait(false);

        return null;
    }

    public static IMessage? HandleFree(DriveClientFreeRequest request)
    {
        Interop.FreeHandle<ProtonDriveClient>(request.ClientHandle);

        return null;
    }
}

File: incubating/client/kt/sdk/src/main/kotlin/me/proton/drive/sdk/internal/InteropProtonDriveClient.kt
package me.proton.drive.sdk.internal

import kotlinx.coroutines.flow.Flow
import kotlinx.coroutines.flow.channelFlow
import me.proton.drive.sdk.Downloader
import me.proton.drive.sdk.FileDownloader
import me.proton.drive.sdk.FileUploader
import me.proton.drive.sdk.LoggerProvider
import me.proton.drive.sdk.LoggerProvider.Level.DEBUG
import me.proton.drive.sdk.LoggerProvider.Level.INFO
import me.proton.drive.sdk.ProtonDriveClient
import me.proton.drive.sdk.SdkNode
import me.proton.drive.sdk.Uploader
import me.proton.drive.sdk.entity.Device
import me.proton.drive.sdk.entity.DeviceType
import me.proton.drive.sdk.entity.DeviceUid
import me.proton.drive.sdk.entity.DriveEvent
import me.proton.drive.sdk.entity.DriveEventId
import me.proton.drive.sdk.entity.FileDownloaderRequest
import me.proton.drive.sdk.entity.FileRevisionUploaderRequest
import me.proton.drive.sdk.entity.FileThumbnail
import me.proton.drive.sdk.entity.FileUploaderRequest
import me.proton.drive.sdk.entity.FolderNode
import me.proton.drive.sdk.entity.Node
import me.proton.drive.sdk.entity.NodeResultPair
import me.proton.drive.sdk.entity.NodeUid
import me.proton.drive.sdk.entity.ScopeId
import me.proton.drive.sdk.entity.ThumbnailType
import me.proton.drive.sdk.extension.toEntity
import me.proton.drive.sdk.extension.toProto
import me.proton.drive.sdk.extension.toTimestamp
import proton.drive.sdk.driveClientCreateDeviceRequest
import proton.drive.sdk.driveClientCreateFolderRequest
import proton.drive.sdk.driveClientDeleteDeviceRequest
import proton.drive.sdk.driveClientDeleteNodesRequest
import proton.drive.sdk.driveClientEmptyTrashRequest
import proton.drive.sdk.driveClientEnumerateDevicesRequest
import proton.drive.sdk.driveClientEnumerateEventsRequest
import proton.drive.sdk.driveClientEnumerateFolderChildrenRequest
import proton.drive.sdk.driveClientEnumerateSharedNodeUidsRequest
import proton.drive.sdk.driveClientEnumerateSharedWithMeNodeUidsRequest
import proton.drive.sdk.driveClientEnumerateThumbnailsRequest
import proton.drive.sdk.driveClientEnumerateTrashRequest
import proton.drive.sdk.driveClientGetAvailableNameRequest
import proton.drive.sdk.driveClientGetMyFilesFolderRequest
import proton.drive.sdk.driveClientGetNodeRequest
import proton.drive.sdk.driveClientLeaveSharedNodeRequest
import proton.drive.sdk.driveClientMoveNodesRequest
import proton.drive.sdk.driveClientRenameDeviceRequest
import proton.drive.sdk.driveClientRenameRequest
import proton.drive.sdk.driveClientRestoreNodesRequest
import proton.drive.sdk.driveClientTrashNodesRequest
import java.time.Instant

internal class InteropProtonDriveClient internal constructor(
    internal val handle: Long,
    private val bridge: JniProtonDriveClient,
) : SdkNode(null), ProtonDriveClient {

    override suspend fun getAvailableName(
        parentFolderUid: NodeUid,
        name: String,
    ): String = cancellationCoroutineScope { source ->
        log(DEBUG, "getAvailableName")
        bridge.getAvailableName(
            driveClientGetAvailableNameRequest {
                this.parentFolderUid = parentFolderUid.value
                this.name = name
                clientHandle = handle
                cancellationTokenSourceHandle = source.handle
            }
        )
    }

    override fun enumerateThumbnails(
        nodeUids: List<NodeUid>,
        type: ThumbnailType,
    ): Flow<FileThumbnail> = channelFlow {
        log(INFO, "enumerateThumbnails(${nodeUids.size}, $type)")
        cancellationCoroutineScope { source ->
            bridge.enumerateThumbnails(
                coroutineScope = this@channelFlow,
                request = driveClientEnumerateThumbnailsRequest {
                    this.fileUids += nodeUids.map { it.value }
                    this.type = type.toProto()
                    clientHandle = handle
                    cancellationTokenSourceHandle = source.handle
                    yieldAction = ProtonDriveSdkNativeClient.getYieldPointer()
                },
                yield = { fileThumbnail ->
                    send(fileThumbnail.toEntity())
                }
            )
        }
    }

    override suspend fun rename(
        nodeUid: NodeUid,
        name: String,
        mediaType: String?,
    ): Unit = cancellationCoroutineScope { source ->
        log(INFO, "rename")
        bridge.rename(
            driveClientRenameRequest {
                this.nodeUid = nodeUid.value
                newName = name
                mediaType?.let {
                    newMediaType = mediaType
                }
                clientHandle = handle
                cancellationTokenSourceHandle = source.handle
            }
        )
    }

    override suspend fun moveNodes(
        nodeUids: List<NodeUid>,
        newParentFolderUid: NodeUid,
    ): List<NodeResultPair> = cancellationCoroutineScope { source ->
        log(INFO, "moveNodes(${nodeUids.size} nodes)")
        bridge.moveNodes(
            driveClientMoveNodesRequest {
                this.nodeUids += nodeUids.map { it.value }
                this.newParentFolderUid = newParentFolderUid.value
                clientHandle = handle
                cancellationTokenSourceHandle = source.handle
            }
        ).toEntity()
    }

    override suspend fun createFolder(
        parentFolderUid: NodeUid,
        name: String,
        lastModificationTime: Instant?,
    ): FolderNode = cancellationCoroutineScope { source ->
        log(INFO, "createFolder")
        bridge.createFolder(
            driveClientCreateFolderRequest {
                this.parentFolderUid = parentFolderUid.value
                folderName = name
                lastModificationTime?.toTimestamp()?.let { this.lastModificationTime = it }
                clientHandle = handle
                cancellationTokenSourceHandle = source.handle
            }
        ).toEntity()
    }

    override suspend fun getMyFilesFolder(): FolderNode = cancellationCoroutineScope { source ->
        log(DEBUG, "getMyFilesFolder")
        bridge.getMyFilesFolder(
            driveClientGetMyFilesFolderRequest {
                clientHandle = handle
                cancellationTokenSourceHandle = source.handle
            }
        ).toEntity()
    }

    override fun enumerateFolderChildrenNodeUids(
        folderUid: NodeUid,
    ): Flow<NodeUid> = channelFlow {
        log(DEBUG, "enumerateFolderChildrenNodeUids")
        cancellationCoroutineScope { source ->
            bridge.enumerateFolderChildrenNodeUids(
                coroutineScope = this@channelFlow,
                request = driveClientEnumerateFolderChildrenRequest {
                    this.folderUid = folderUid.value
                    clientHandle = handle
                    cancellationTokenSourceHandle = source.handle
                    yieldAction = ProtonDriveSdkNativeClient.getYieldPointer()
                },
                yield = { nodeUid ->
                    send(NodeUid(nodeUid.value))
                }
            )
        }
    }

    override fun enumerateEvents(
        scopeId: ScopeId,
        cursorEventId: DriveEventId?,
    ): Flow<DriveEvent> = channelFlow {
        log(DEBUG, "enumerateEvents")
        cancellationCoroutineScope { source ->
            bridge.enumerateEvents(
                coroutineScope = this@channelFlow,
                request = driveClientEnumerateEventsRequest {
                    this.treeEventScopeId = scopeId.id
                    cursorEventId?.let { this.cursorEventId = it.value }
                    clientHandle = handle
                    cancellationTokenSourceHandle = source.handle
                    yieldAction = ProtonDriveSdkNativeClient.getYieldPointer()
                },
                yield = { event ->
                    send(event.toEntity())
                }
            )
        }
    }

    override suspend fun getNode(
        nodeUid: NodeUid,
    ): Node? = cancellationCoroutineScope { source ->
        log(DEBUG, "getNode")
        bridge.getNode(
            driveClientGetNodeRequest {
                this.nodeUid = nodeUid.value
                clientHandle = handle
                cancellationTokenSourceHandle = source.handle
            }
        )?.toEntity()
    }

    override fun trashNodes(
        nodeUids: List<NodeUid>,
    ): Flow<NodeResultPair> = channelFlow {
        log(INFO, "trashNodes(${nodeUids.size} nodes)")
        cancellationCoroutineScope { source ->
            bridge.trashNodes(
                coroutineScope = this@channelFlow,
                driveClientTrashNodesRequest {
                    this.nodeUids += nodeUids.map { it.value }
                    clientHandle = handle
                    cancellationTokenSourceHandle = source.handle
                    yieldAction = ProtonDriveSdkNativeClient.getYieldPointer()
                },
                yield = { pair ->
                    send(pair.toEntity())
                }
            )
        }
    }

    override fun deleteNodes(
        nodeUids: List<NodeUid>,
    ): Flow<NodeResultPair> = channelFlow {
        log(INFO, "deleteNodes(${nodeUids.size} nodes)")
        cancellationCoroutineScope { source ->
            bridge.deleteNodes(
                coroutineScope = this@channelFlow,
                driveClientDeleteNodesRequest {
                    this.nodeUids += nodeUids.map { it.value }
                    clientHandle = handle
                    cancellationTokenSourceHandle = source.handle
                    yieldAction = ProtonDriveSdkNativeClient.getYieldPointer()
                },
                yield = { pair ->
                    send(pair.toEntity())
                }
            )
        }
    }

    override fun restoreNodes(
        nodeUids: List<NodeUid>,
    ): Flow<NodeResultPair> = channelFlow {
        log(INFO, "restoreNodes(${nodeUids.size} nodes)")
        cancellationCoroutineScope { source ->
            bridge.restoreNodes(
                coroutineScope = this@channelFlow,
                driveClientRestoreNodesRequest {
                    this.nodeUids += nodeUids.map { it.value }
                    clientHandle = handle
                    cancellationTokenSourceHandle = source.handle
                    yieldAction = ProtonDriveSdkNativeClient.getYieldPointer()
                },
                yield = { pair ->
                    send(pair.toEntity())
                }
            )
        }
    }

    override fun enumerateTrashNodeUids(): Flow<NodeUid> = channelFlow {
        log(DEBUG, "enumerateTrashNodeUids")
        cancellationCoroutineScope { source ->
            bridge.enumerateTrash(
                coroutineScope = this@channelFlow,
                request = driveClientEnumerateTrashRequest {
                    clientHandle = handle
                    cancellationTokenSourceHandle = source.handle
                    yieldAction = ProtonDriveSdkNativeClient.getYieldPointer()
                },
                yield = { nodeUid ->
                    send(NodeUid(nodeUid.value))
                }
            )
        }
    }

    override suspend fun emptyTrash(): Unit = cancellationCoroutineScope { source ->
        log(INFO, "emptyTrash")
        bridge.emptyTrash(
            driveClientEmptyTrashRequest {
                clientHandle = handle
                cancellationTokenSourceHandle = source.handle
            }
        )
    }

    override fun enumerateSharedNodeUids(): Flow<NodeUid> = channelFlow {
        log(DEBUG, "enumerateSharedNodeUids")
        cancellationCoroutineScope { source ->
            bridge.enumerateSharedNodeUids(
                coroutineScope = this@channelFlow,
                request = driveClientEnumerateSharedNodeUidsRequest {
                    clientHandle = handle
                    cancellationTokenSourceHandle = source.handle
                    yieldAction = ProtonDriveSdkNativeClient.getYieldPointer()
                },
                yield = { nodeUid ->
                    send(NodeUid(nodeUid.value))
                }
            )
        }
    }

    override fun enumerateSharedWithMeNodeUids(): Flow<NodeUid> = channelFlow {
        log(DEBUG, "enumerateSharedWithMeNodeUids")
        cancellationCoroutineScope { source ->
            bridge.enumerateSharedWithMeNodeUids(
                coroutineScope = this@channelFlow,
                request = driveClientEnumerateSharedWithMeNodeUidsRequest {
                    clientHandle = handle
                    cancellationTokenSourceHandle = source.handle
                    yieldAction = ProtonDriveSdkNativeClient.getYieldPointer()
                },
                yield = { nodeUid ->
                    send(NodeUid(nodeUid.value))
                }
            )
        }
    }

    override suspend fun leaveSharedNode(
        nodeUid: NodeUid,
    ): Unit = cancellationCoroutineScope { source ->
        log(INFO, "leaveSharedNode")
        bridge.leaveSharedNode(
            driveClientLeaveSharedNodeRequest {
                this.nodeUid = nodeUid.value
                clientHandle = handle
                cancellationTokenSourceHandle = source.handle
            }
        )
    }

    override fun enumerateDevices(): Flow<Device> = channelFlow {
        log(DEBUG, "enumerateDevices")
        cancellationCoroutineScope { source ->
            bridge.enumerateDevices(
                coroutineScope = this@channelFlow,
                request = driveClientEnumerateDevicesRequest {
                    clientHandle = handle
                    cancellationTokenSourceHandle = source.handle
                    yieldAction = ProtonDriveSdkNativeClient.getYieldPointer()
                },
                yield = { device ->
                    send(device.toEntity())
                }
            )
        }
    }

    override suspend fun createDevice(
        name: String,
        type: DeviceType,
    ): Device = cancellationCoroutineScope { source ->
        log(INFO, "createDevice")
        bridge.createDevice(
            driveClientCreateDeviceRequest {
                this.name = name
                deviceType = type.toProto()
                clientHandle = handle
                cancellationTokenSourceHandle = source.handle
            }
        ).toEntity()
    }

    override suspend fun renameDevice(
        deviceUid: DeviceUid,
        name: String,
    ): Device = cancellationCoroutineScope { source ->
        log(INFO, "renameDevice")
        bridge.renameDevice(
            driveClientRenameDeviceRequest {
                this.deviceUid = deviceUid.value
                this.name = name
                clientHandle = handle
                cancellationTokenSourceHandle = source.handle
            }
        ).toEntity()
    }

    override suspend fun deleteDevice(
        deviceUid: DeviceUid,
    ): Unit = cancellationCoroutineScope { source ->
        log(INFO, "deleteDevice")
        bridge.deleteDevice(
            driveClientDeleteDeviceRequest {
                this.deviceUid = deviceUid.value
                clientHandle = handle
                cancellationTokenSourceHandle = source.handle
            }
        )
    }

    override suspend fun downloader(
        request: FileDownloaderRequest,
    ): Downloader {
        log(INFO, "downloader")
        return cancellationCoroutineScope { source ->
            factory(JniFileDownloader()) {
                FileDownloader(
                    client = this@InteropProtonDriveClient,
                    handle = getFileDownloader(
                        clientHandle = handle,
                        cancellationTokenSourceHandle = source.handle,
                        request = request,
                    ),
                    bridge = this,
                    cancellationTokenSource = source,
                )
            }
        }
    }

    override suspend fun uploader(
        request: FileUploaderRequest,
    ): Uploader {
        log(INFO, "fileUploader")
        return cancellationCoroutineScope { source ->
            JniFileUploader().run {
                FileUploader(
                    client = this@InteropProtonDriveClient,
                    handle = getFileUploader(
                        clientHandle = handle,
                        cancellationTokenSourceHandle = source.handle,
                        request = request,
                    ),
                    bridge = this,
                    cancellationTokenSource = source,
                )
            }
        }
    }

    override suspend fun uploader(
        request: FileRevisionUploaderRequest,
    ): Uploader {
        log(INFO, "fileRevisionUploader")
        return cancellationCoroutineScope { source ->
            JniFileUploader().run {
                FileUploader(
                    client = this@InteropProtonDriveClient,
                    handle = getFileRevisionUploader(
                        clientHandle = handle,
                        cancellationTokenSourceHandle = source.handle,
                        request = request,
                    ),
                    bridge = this,
                    cancellationTokenSource = source,
                )
            }
        }
    }

    override fun close() {
        log(DEBUG, "close")
        bridge.free(handle)
        super.close()
    }

    private fun log(level: LoggerProvider.Level, message: String) {
        bridge.clientLogger(level, "DriveClient(${handle.toLogId()}) $message")
    }
}
